crates/iceberg-catalog/src/api/iceberg/mod.rs

```diff
@@ -67,46 +68,135 @@ pub mod v1 {
 
-    impl<T, Z> PaginatedTabulars<T, Z>
+    impl<T, V> PaginatedMapping<T, V>
     where
-        T: std::hash::Hash + Eq + Debug,
-        Z: Debug,
+        T: std::hash::Hash + Eq + Debug + Clone + 'static,
+        V: Debug + 'static,
     {
+        pub(crate) fn map<
+            NewKey: std::hash::Hash + Eq + Debug + Clone + 'static,
+            NewVal: Debug + 'static,
+        >(
+            self,
+            key_map: impl Fn(T) -> Result<NewKey>,
+            value_map: impl Fn(V) -> Result<NewVal>,
+        ) -> Result<PaginatedMapping<NewKey, NewVal>> {
+            let mut new_mapping = PaginatedMapping::with_capacity(self.len());
+            for (key, value, token) in self.into_iter_with_page_tokens() {
+                let k = key_map(key)?;
+                new_mapping.insert(k, value_map(value)?, token);
+            }
+            Ok(new_mapping)
+        }
+
+        #[must_use]
+        pub(crate) fn with_capacity(capacity: usize) -> Self {
+            Self {
+                entities: HashMap::with_capacity(capacity),
+                next_page_tokens: Vec::with_capacity(capacity),
+                ordering: Vec::with_capacity(capacity),
+            }
+        }
+
+        pub(crate) fn insert(&mut self, key: T, value: V, next_page_token: String) {
+            // TODO: should this become a result instead of a silent overwrite?
+            if self.entities.insert(key.clone(), value).is_some() {
+                let position = self
+                    .ordering
+                    .iter()
+                    .find_position(|item| **item == key)
+                    .map(|(idx, _)| idx);
+                if let Some(idx) = position {
+                    self.ordering.remove(idx);
+                    self.next_page_tokens.remove(idx);
+                }
+            };
+            self.ordering.push(key);
+            self.next_page_tokens.push(next_page_token);
+        }
+
         #[must_use]
         pub fn len(&self) -> usize {
-            self.tabulars.len()
+            self.entities.len()
         }
 
         #[must_use]
         pub fn is_empty(&self) -> bool {
-            self.tabulars.is_empty()
+            self.entities.is_empty()
+        }
+
+        pub fn get(&self, key: &T) -> Option<&V> {
+            self.entities.get(key)
+        }
+
+        #[cfg(test)]
+        pub fn remove(&mut self, key: &T) -> Option<V> {
+            let (idx, _) = self.ordering.iter().find_position(|item| **item == *key)?;
+            self.ordering.remove(idx);
+            self.next_page_tokens.remove(idx);
+            self.entities.remove(key)
+        }
+
+        #[allow(clippy::missing_panics_doc)]
+        pub fn into_iter_with_page_tokens(mut self) -> impl Iterator<Item = (T, V, String)> {
+            self.ordering
+                .into_iter()
+                .zip(self.next_page_tokens)
+                // we can unwrap here since the only way of adding items is via insert which ensures that every
```

Comment:
I think this can lead to panics currently if the insert "silent override" hits.
In this case we still push ordering & next_page_token, which can contain duplicates as they are Vecs.
While `.remove` the first item works, the second panics

Reply:
The vecs are guaranteed to have unique items:

We start with an empty collection, the only way to add items is via insert.

1. insert item A: 1 item in hashmap, ordering and next page token
2. insert item B: 2 items in hashmap, ordering and next page token
3. insert item B: we insert into hashmap, hashmap returns a V, we find the index of K in both vecs and remove them before pushing again

since there is no other way of adding items, we know that the vecs have a uniqueness guarantee.

I've added some tests, could you check if your case is covered?